lib/rubocop/cop/rails/index_names.rb

```diff
@@ -0,0 +1,114 @@
+# typed: false
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # Checks for custom index names in migrations.
+      #
+      # @example
+      #   # bad
+      #   class ExampleMigration < ActiveRecord::Migration[7.1]
+      #     def change
+      #       change_table :users do |t|
+      #         t.index [:email], name: 'index_custom_name'
+      #       end
+      #     end
+      #   end
+      #
+      #   # good
+      #   class ExampleMigration < ActiveRecord::Migration[7.1]
+      #     def change
+      #       create_table :users do |t|
+      #         t.index [:email]
+      #       end
+      #     end
+      #   end
+      class IndexNames < Base
+        include MigrationsHelper
+        include RangeHelp
+        extend AutoCorrector
+        extend TargetRailsVersion
+
+        minimum_target_rails_version 7.1
+
+        MSG = 'Avoid specifying a custom name for common indexes. Let Rails handle the index name automatically.'
+        RESTRICT_ON_SEND = %i[index add_index].freeze
+        # Keys that do not justify a duplicative index or a custom name:
+        # unique, length, order, opclass, nulls_not_distinct
```

Comment:
My reasoning is that even if you were rolling out a replacement index adding one of these properties, the rubocop: disable statement would serve as a valid reminder to come back and remove the superseded index. 

If I misunderstood one of these and there existss a good reason to keep 2 indexes (both with and without this property) on identical columns for the long term, please let me know and I will update accordingly by moving it to the allowlist.

Reply:
https://github.com/bensheldon/good_job/pull/1213 introduces the same index with different `order`ing for good_job so there definitly can be use-cases where it is appropriate. But it of course depends on each use-case. Personally I'm not qualified enough to say it one way or the other (except where I observed it like above). There may be valid use-cases for the other options as well

Reply:
Thanks! I'm going to move `order` and `opclass` into the allow list to make this rollout smoother. 

I think we can make a strong case that no one needs the same keys to be both a unique and a non-unique index at the same time in the long term. Similarly, we should not need the same columns to have indexes with different lengths specified. Adding `nulls_not_distinct ` also looks like it would result in duplication only during a migration, as it is [strictly stricter than the normal unique index ](https://www.postgresql.org/about/featurematrix/detail/392/). With this option, `NULL` is checked for uniqueness just like any other field.